simba/schedule.py

```diff
@@ -285,10 +301,170 @@ def assign_vehicles(self):
+                               for vehicle_type, charging_types in self.vehicle_types.items()
+                               for charging_type in charging_types.keys()}
+
+        charge_levels = {station.get(param) for name, station in self.stations.items()
+                         for param in ["cs_power_deps_oppb", "cs_power_deps_depb"]
+                         if station["type"] == "deps"}
+        # Remove None as charge level
+        charge_levels = charge_levels.difference([None])
+        # Add default values from arguments
+        charge_levels.add(self.cs_power_deps_depb)
+        charge_levels.add(self.cs_power_deps_oppb)
+
+        # Calculates numeric charge curves for each charge_level. The charge levels might clip the
+        # charge curve of the vehicle.
+        charge_curves = self.get_charge_curves(charge_levels, initial_soc, time_step_min=1)
+
+        rotations = sorted(self.rotations.values(), key=lambda rot: rot.departure_time)
+        for k, rot in enumerate(rotations):
+            # find vehicles that have completed their rotation
+            while rotations_in_progress:
+                r = rotations_in_progress.pop(0)
+                if rot.departure_time > r.arrival_time:
+                    all_standing_vehicles.append((r.vehicle_id, r.arrival_name))
+                else:
+                    rotations_in_progress.insert(0, r)
+                    break
+
+            # find standing vehicle for rotation if exists
+            # else generate new vehicle id
+            vt = rot.vehicle_type
+            ct = rot.charging_type
+            vt_ct = f"{vt}_{ct}"
+
+            station_is_electrified = True
+            try:
+                self.stations[rot.departure_name]
+            except KeyError:
+                logging.warning(f"Rotation {rot.id} ends at a non electrified station.")
+                station_is_electrified = False
+
+            # filter vehicles of the right vehicle type and right location
+            standing_vehicles = list(filter(lambda x: vt_ct in x[0] * (x[1] == rot.departure_name),
+                                            all_standing_vehicles))
+
+            def partial_function(v_id_deps: tuple):
+                """ get the partial function of soc_at_departure_time
+
+                :param v_id_deps: vehicle_id and depot name
+                :type v_id_deps: tuple
+                :returns: partial function
+                :rtype: Callable
+                """
+                return soc_at_departure_time(v_id_deps, departure_time, vehicle_data, self.stations,
+                                             charge_curves, args)
+
+            # Add the expected soc at departure time, if greedy charging would be applied
+            socs = list(map(lambda x: partial_function(x), standing_vehicles))
+            # Pair with socs
+            standing_vehicles_w_soc = [(*v, k) for v, k in zip(standing_vehicles, socs)]
```

Comment:
dangerous to use k here, is already an index variable -> rename k to rot_idx in for-loop l.344